emails/tests/views_tests.py

```diff
@@ -594,6 +596,30 @@ def test_successful_email_in_s3_deleted(
         assert response.status_code == 200
         assert response.content == b"Email relayed"
 
+    @override_settings(STATSD_ENABLED=True)
+    @patch("emails.views.ses_relay_email")
+    @patch("emails.views._get_text_html_attachments")
+    @patch("emails.views.remove_message_from_s3")
+    def test_dmarc_failure_s3_deleted(
+        self,
+        mocked_message_removed,
+        mocked_get_text_html,
+        mocked_relay_email,
+    ):
+        """A message with a failing DMARC and a "reject" policy is rejected."""
+        mocked_get_text_html.return_value = ("text_content", None, ["attachments"])
+        mocked_relay_email.side_effect = Exception("Not called")
```

Comment:
My only complaint is the long string turns it into a 3-line statement. Which is OK, explicit is better than implicit.

Other code uses `Exception("Should not be called.")`, slightly longer but still a one-liner:

https://github.com/mozilla/fx-private-relay/blob/16f964ccc586710064e14c29bf9b6e763ff08e2c/emails/tests/views_tests.py#L864-L865

We could also make a global that all tests can use:

```python
# Use as a mocked function's side_effect when the mock should not be called.
FAIL_TEST_IF_CALLED = Exception("This function should not have been called.")

...

mocked_relay_email.side_effect = FAIL_TEST_IF_CALLED
```

Any preference?

Reply:
I went with `FAIL_TEST_IF_CALLED = Exception("This function should not have been called.")`

Reply:
I went with `FAIL_TEST_IF_CALLED = Exception("This function should not have been called.")`

Reply:
What does a test failure look like with this? Does it have the function name somewhere in the stacktrace?

Reply:
I've added a failing test to https://github.com/mozilla/fx-private-relay/pull/1993 to demonstrate. Output:

```
self = <emails.tests.views_tests.SNSNotificationTest testMethod=test_single_recipient_sns_notification>
mock_ses_relay_email = <MagicMock name='ses_relay_email' id='140443475331296'>

    @patch("emails.views.ses_relay_email")
    def test_single_recipient_sns_notification(self, mock_ses_relay_email):
        mock_ses_relay_email.return_value = HttpResponse(
            "Successfully relayed emails", status=200
        )
        # Demo FAIL_TEST_IF_CALLED
        mock_ses_relay_email.side_effect = FAIL_TEST_IF_CALLED
>       _sns_notification(EMAIL_SNS_BODIES["single_recipient"])

emails/tests/views_tests.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
emails/views.py:300: in _sns_notification
    response = _sns_message(message_json)
emails/views.py:518: in _sns_message
    response = ses_relay_email(
/usr/local/lib/python3.9/unittest/mock.py:1092: in __call__
    return self._mock_call(*args, **kwargs)
/usr/local/lib/python3.9/unittest/mock.py:1096: in _mock_call
    return self._execute_mock_call(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <MagicMock name='ses_relay_email' id='140443475331296'>
args = ('=?utf-8?q?=22fxastage=40protonmail=2Ecom_=5Bvia_Relay=5D=22?= <relay@127.0.0.1>', '', 'localized email header + foot...im=pass header.i=@protonmail.com; dmarc=pass header.from=protonmail.com;'}, ...], 'headersTruncated': False, ...}, ...)
kwargs = {}, effect = Exception('This function should not have been called.')

    def _execute_mock_call(self, /, *args, **kwargs):
        # separate from _increment_mock_call so that awaited functions are
        # executed separately from their call, also AsyncMock overrides this method
    
        effect = self.side_effect
        if effect is not None:
            if _is_exception(effect):
>               raise effect
E               Exception: This function should not have been called.

/usr/local/lib/python3.9/unittest/mock.py:1151: Exception
```